products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,345 @@
+    }
+    
+    /// Get the key for a given key identifier (might be unique)
+    public func getKey(id: KeyID) -> KeyType {
+        assert(result != nil, "Can't get key without fetching the data first.")
+        
+        var data = llb_data_t()
+        withUnsafeMutablePointer(to: &data) { ptr in
+            llb_database_result_keys_get_key_for_id(self.result, id, ptr)
+        }
+        defer { data.data?.deallocate() }
+        return stringFromData(data)
+    }
+    
+    deinit {
+        llb_database_destroy_result_keys(result)
+    }
+}
+
+extension BuildDBFetchKeysResult: Collection {
+    public typealias Index = KeyID
+    public typealias Element = (keyID: KeyID, key: KeyType)
+
+    public var startIndex: Index {
+        return 1 as UInt64
+    }
+    
+    public var endIndex: Index {
+        return UInt64(self.count) + startIndex
+    }
+    
+    public subscript(index: Index) -> Iterator.Element {
+        guard (startIndex..<endIndex).contains(index) else {
+            fatalError("Index \(index) is out of bounds (\(startIndex)..<\(endIndex))")
+        }
+        return (index, getKey(id: index))
+    }
+    
+    public func index(after i: KeyID) -> KeyID {
+        return i + 1
+    }
+}
+
+extension BuildDBFetchKeysResult: BuildDBDelegate {}
+
+/// Defines the result of a built task
+public class RuleResult {
+    /// Reference to the C API result instance (for desctruction)
+    private var result: llb_database_result_t
+    /// The value of the result
+    public let value: Data
+    /// Signature of the node that generated the result
+    public let signature: UInt64
+    /// The build iteration this result was computed at
+    public let computedAt: UInt64
+    /// The build iteration this result was built at
+    public let builtAt: UInt64
+    /// A list of the dependencies of the computed task, use the database's allKeys to check for their key
+    public let dependencies: [KeyID]
+    
+    fileprivate init(result: llb_database_result_t) {
+        self.result = result
+        value = Data(bytes: result.value.data, count: Int(result.value.length))
+        signature = result.signature
+        computedAt = result.computed_at
+        builtAt = result.built_at
+        
+        dependencies = UnsafeBufferPointer(start: result.dependencies, count: Int(result.dependencies_count)).map { KeyID($0) }
+    }
+    
+    deinit {
+        withUnsafeMutablePointer(to: &result) { ptr in
+            llb_database_destroy_result(ptr)
+        }
+    }
+    
+    /// Fetch result information about all dependencies via the given database
+    public func fetchDependencies(database: Database.BuildDB) throws -> [RuleResult] {
+        var result = [RuleResult]()
+        
+        for dep in dependencies {
+            guard let res = try database.lookupRuleResult(keyID: dep) else {
+                continue
+            }
+            result.append(res)
+        }
+        
+        return result
+    }
```

Comment:
This function is as simple as: `return try dependencies.compactMap{ database.lookupRuleResult(keyID: $0) }` isn't it?

I would recommend not exposing this in the API. It isn't particularly efficient, and the client can easily write the same code, which will make it obvious that they should probably be building their own cache.

If we want conveniences like this, I'd suggest building them in a separate layer separate from the raw bindings, and I'd leave them out of the initial PR.